src/function.ts

```diff
@@ -61,28 +125,539 @@ export class Function<P, O> implements Integration {
+        local: {
+          tryBundle(outdir: string) {
+            fs.writeFileSync(path.resolve(outdir, "index.js"), resultText);
+            return true;
+          },
+        },
+      },
+    });
+
+    if (!(scope instanceof aws_lambda.Function)) {
+      throw new Error(
+        "CallbackLambdaCode can only be used on aws_lambda.Function"
+      );
+    }
+
+    tokenContext.forEach((t) => scope.addEnvironment(t.env, t.token));
+
+    const funcResource = scope.node.findChild(
+      "Resource"
+    ) as aws_lambda.CfnFunction;
+
+    const codeConfig = asset.bind(scope);
+
+    funcResource.code = {
+      s3Bucket: codeConfig.s3Location?.bucketName,
+      s3Key: codeConfig.s3Location?.objectKey,
+      s3ObjectVersion: codeConfig.s3Location?.objectVersion,
+      zipFile: codeConfig.inlineCode,
+      imageUri: codeConfig.image?.imageUri,
+    };
+
+    asset.bindToResource(funcResource);
+  }
+}
+
+/**
+ * Interface to consume to add an Integration to Native Lambda Functions.
+ *
+ * ```ts
+ * new Function(this, 'func', () => {
+ *    mySpecialIntegration()
+ * })
+ *
+ * const mySpecialIntegration = makeIntegration<() => void>({
+ *    native: {...} // an instance of NativeIntegration
+ * })
+ * ```
+ */
+export interface NativeIntegration<F extends AnyFunction> {
+  /**
+   * Called by any {@link Function} that will invoke this integration during CDK Synthesis.
+   * Add permissions, create connecting resources, validate.
+   *
+   * @param context - The function invoking this function.
+   * @param args - The functionless encoded AST form of the arguments passed to the integration.
+   */
+  bind: (context: Function<any, any>, args: Expr[]) => void;
+  /**
+   * @param args The arguments passed to the integration function by the user.
+   * @param preWarmContext contains singleton instances of client and other objects initialized outside of the native
+   *                       function handler.
+   */
+  call: (
+    args: Parameters<F>,
+    preWarmContext: NativePreWarmContext
+  ) => Promise<ReturnType<F>>;
+  /**
+   * Method called outside of the handler to initialize things like the PreWarmContext
+   */
+  preWarm?: (preWarmContext: NativePreWarmContext) => void;
+}
+
+export type ClientName =
+  | "LAMBDA"
+  | "EVENT_BRIDGE"
+  | "STEP_FUNCTIONS"
+  | "DYNAMO";
+
+interface PrewarmProps {
+  clientConfigRetriever?: FunctionProps["clientConfigRetriever"];
+}
+
+export interface PrewarmClientInitializer<T, O> {
+  key: T;
+  init: (key: string, props?: PrewarmProps) => O;
+}
+
+/**
+ * Known, shared clients to use.
+ *
+ * Any object can be used by using the {@link PrewarmClientInitializer} interface directly.
+ *
+ * ```ts
+ * context.getOrInit({
+ *   key: 'customClient',
+ *   init: () => new anyClient()
+ * })
+ * ```
+ */
+export const PrewarmClients = {
+  LAMBDA: {
+    key: "LAMBDA",
+    init: (key, props) =>
+      new (require("aws-sdk").Lambda)(props?.clientConfigRetriever?.(key)),
```

Comment:
`aws-sdk` is ideal for bundling because it's included in the lambda function by default. Do we know if it has any performance implications though? Does V2 use the optimized C code that v3 does?

Reply:
There may be some performance improvements using v3 in a function: https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/

https://github.com/functionless/functionless/issues/174